# DEMOGRAPHIC DATA ANALYZER

This is me 'completing' a project designed by freecodecamp, except using Jupyter. 

In [2]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import matplotlib
import os

In [3]:
path = r'/Users/bryanleavelle/DATA PROJECTS/freecodecamp-demographix'

In [4]:
df = pd.read_csv(os.path.join(path, 'adult.data.csv'))

# FIRST CHECK AND CLEAN

In [5]:
df.shape


(32561, 15)

In [6]:
df.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [7]:
df.dtypes

age                int64
workclass         object
fnlwgt             int64
education         object
education-num      int64
marital-status    object
occupation        object
relationship      object
race              object
sex               object
capital-gain       int64
capital-loss       int64
hours-per-week     int64
native-country    object
salary            object
dtype: object

In [8]:
# check for missing values
df.isnull().sum()

age               0
workclass         0
fnlwgt            0
education         0
education-num     0
marital-status    0
occupation        0
relationship      0
race              0
sex               0
capital-gain      0
capital-loss      0
hours-per-week    0
native-country    0
salary            0
dtype: int64

In [9]:
# check for duplicates

df_dups = df[df.duplicated()]

In [10]:
df_dups.shape

(24, 15)

Well thats interesting. Get 'em outa here.

In [11]:
df = df.drop_duplicates()

In [12]:
df_dups = df[df.duplicated()]

In [13]:
df_dups

Empty DataFrame
Columns: [age, workclass, fnlwgt, education, education-num, marital-status, occupation, relationship, race, sex, capital-gain, capital-loss, hours-per-week, native-country, salary]
Index: []

lookin good

# QUESTION 1: 
## How many people of each race are represented in this dataset? 

In [14]:
ser_race =df['race'].value_counts()

In [15]:
ser_race

White                 27795
Black                  3122
Asian-Pac-Islander     1038
Amer-Indian-Eskimo      311
Other                   271
Name: race, dtype: int64

boom, see above. Next.

# QUESTION 2: 
## What is the average age of men?

In [16]:
#group by sex and calculate the average

df.groupby('sex', as_index = False).age.mean().round(decimals = 1)

sex   age
0  Female  36.9
1    Male  39.4

Average age for men is 39.43 years old. Next.

# QUESTION 3: 

## What is the percentage of people who have a Bachelor's degree?

In [17]:
#simple normalized value counts

df['education'].value_counts(normalize = True).round(decimals = 3)

HS-grad         0.323
Some-college    0.224
Bachelors       0.165
Masters         0.053
Assoc-voc       0.042
11th            0.036
Assoc-acdm      0.033
10th            0.029
7th-8th         0.020
Prof-school     0.018
9th             0.016
12th            0.013
Doctorate       0.013
5th-6th         0.010
1st-4th         0.005
Preschool       0.002
Name: education, dtype: float64

~16.5 percent of people have a Bachelor's degree. Nnnnnnnnext!

# QUESTION 4: 
## What percentage of people with advanced education (Bachelors, Masters, or Doctorate) make more than 50K?

Lets start off by making a new column 'advanced_ed' that flags each entry with a 'yes' or 'no'


In [18]:
# use .loc to find people with advanced degrees, and plop a 'yes' into the new 'advanced_ed' column

df.loc[(df['education'] == 'Bachelors'), 'advanced_ed'] = 'yes'
df.loc[(df['education'] == 'Masters'), 'advanced_ed'] = 'yes'
df.loc[(df['education'] == 'Doctorate'), 'advanced_ed'] = 'yes'

In [19]:
# change NaN from previous code into 'no'

df['advanced_ed'] = df['advanced_ed'].replace(np.nan, 'no')

In [20]:
#check work

df['advanced_ed'].value_counts()

no     25049
yes     7488
Name: advanced_ed, dtype: int64

Nice

In [21]:
# group by new flag and calculate the proprotions of salary groups 

df.groupby('advanced_ed', as_index=False).salary.value_counts(normalize =True ).round(decimals = 3)

advanced_ed salary  proportion
0          no  <=50K       0.826
1          no   >50K       0.174
2         yes  <=50K       0.534
3         yes   >50K       0.466

~46.6 percent of people with advanced education make over 50k a year. 

# QUESTION 5:  
## What percentage of people without advanced education make more than 50K?

Freebie! From above: ~17.4% of people without advanced education make more than 50k per year.

# QUESTION 6: 
## What is the minimum number of hours a person works per week?

In [22]:
# simple describe of the hours-per-week column

df['hours-per-week'].describe().round(decimals = 2)

count    32537.00
mean        40.44
std         12.35
min          1.00
25%         40.00
50%         40.00
75%         45.00
max         99.00
Name: hours-per-week, dtype: float64

Minimum is 1 hour per week.

# QUESTION 7: 
## What percentage of the people who work the minimum number of hours per week have a salary of more than 50K?

omg i hope this answer is not horrifying.

In [25]:
# group by hours per week and then find the proportions of salary categories for that group and save as new df

df_work = df.groupby('hours-per-week', as_index=False).salary.value_counts(normalize =True ).round(decimals = 3)

#filter the top two results (because they will be the minimum and will show up first and there will be two of them)

df_work.filter(items = [0,1], axis=0)


hours-per-week salary  proportion
0               1  <=50K         0.9
1               1   >50K         0.1

ten percent of people who work one hour per week make over 50k. Yikes.

# QUESTION 8:
## What country has the highest percentage of people that earn >50K and what is that percentage?

In [113]:
#new dataframe from a groupby that contains countries and salary value counts.

df2 = df.groupby('native-country', as_index=False).salary.value_counts(normalize =True )

In [115]:
#check new df

df2

native-country salary  proportion
0               ?  <=50K    0.749141
1               ?   >50K    0.250859
2        Cambodia  <=50K    0.631579
3        Cambodia   >50K    0.368421
4          Canada  <=50K    0.677686
..            ...    ...         ...
77  United-States   >50K    0.245910
78        Vietnam  <=50K    0.925373
79        Vietnam   >50K    0.074627
80     Yugoslavia  <=50K    0.625000
81     Yugoslavia   >50K    0.375000

[82 rows x 3 columns]

In [116]:
# remove <=50K from df

df2 = df2[df2['salary'] == '>50K']

In [128]:
# sort by hightest proportion value

df2.sort_values(by = 'proportion', ascending=False).round(decimals = 3)

native-country salary  proportion
40                Iran   >50K       0.419
21              France   >50K       0.414
38               India   >50K       0.400
71              Taiwan   >50K       0.392
48               Japan   >50K       0.387
81          Yugoslavia   >50K       0.375
3             Cambodia   >50K       0.368
44               Italy   >50K       0.342
19             England   >50K       0.333
5               Canada   >50K       0.322
23             Germany   >50K       0.321
59         Philippines   >50K       0.308
34                Hong   >50K       0.300
25              Greece   >50K       0.276
7                China   >50K       0.267
11                Cuba   >50K       0.263
1                    ?   >50K       0.251
67            Scotland   >50K       0.250
77       United-States   >50K       0.246
36             Hungary   >50K       0.231
42             Ireland   >50K       0.208
69               South   >50K       0.200
61              Poland   >50K       0.200
73            Thailand   >50K       0.167
15             Ecuador   >50K       0.143
46             Jamaica   >50K       0.123
50                Laos   >50K       0.111
63            Portugal   >50K       0.108
65         Puerto-Rico   >50K       0.105
75     Trinadad&Tobago   >50K       0.105
29               Haiti   >50K       0.091
17         El-Salvador   >50K       0.085
32            Honduras   >50K       0.077
79             Vietnam   >50K       0.075
57                Peru   >50K       0.065
54           Nicaragua   >50K       0.059
52              Mexico   >50K       0.052
27           Guatemala   >50K       0.048
9             Columbia   >50K       0.034
13  Dominican-Republic   >50K       0.029

IRAN has the highest proportion of people that make > 50K, and that proportion is ~41.9%

# FIN